### ISSUE TITLE
Weirdness in Either projections

### ISSUE THREAD

**Issue by Blaisorblade** [scala/bug#10285]
I can't wrap my head around Either.right.filter. Compare these two variants:
```
scala> Right(1).right.filter(x => true).map(x => x)
res4: Option[scala.util.Either[Nothing,Int]] = Some(Right(1))

scala> Right(1).right.map(x => x)
res5: scala.util.Either[Nothing,Int] = Right(1)
```
So `filter` suggests `right` gives an `Option[Either[A, B]]`, while `map` suggests `right` gives an `Either[A, B]`. Weird, huh?

If it had a `withFilter` aliasing `filter`, you could write the above also as for comprehensions, as you could till 2.11:
```
scala> for (x <- Right(1).right) yield x
res0: scala.util.Either[Nothing,Int] = Right(1)

scala> for (x <- Right(1).right; if true) yield x
<console>:12: error: value withFilter is not a member of scala.util.Either.RightProjection[Nothing,Int]
       for (x <- Right(1).right; if true) yield x
                          ^
```
The only previous semi-related discussion was in https://github.com/scala/bug/issues/5793.

This came up while re-debugging https://github.com/scala/bug/issues/5589.

**Comment by refried** [comment]
What should `Right(1).right.filter(x => false)` return?

**Comment by Blaisorblade** [comment]
Yeah, I should have been clearer. I know that's a good question. I'm just saying that since you can't answer it within the "standard" filter interface, then it'd be better use another name instead of filter. Waiting for that, we could add a warning.

**Comment by SethTisue** [comment]
PR: scala/scala#7379

**Comment by SethTisue** [comment]
@Nasadorian if you comment here after accepting my invite to https://github.com/orgs/scala/teams/contributors, I'll assign you this ticket

**Comment by SethTisue** [comment]
P.S. also please confirm you aren't the devil <img width="453" alt="screen shot 2018-11-09 at 6 30 46 pm" src="https://user-images.githubusercontent.com/161079/48278386-a106eb00-e44d-11e8-9f37-8267ceb199bb.png">

**Comment by nasadorian** [comment]
@SethTisue That is so epic I will stop writing code just to keep it that way
